midend/interpreter.cpp

```diff
@@ -1105,11 +1126,27 @@ void ExpressionEvaluator::postorder(const IR::MethodCallExpression* expression)
         auto bim = mi->to<BuiltInMethod>();
         auto base = get(bim->appliedTo);
         cstring name = bim->name.name;
+        // This code is necessary to find the parent and to check whether the parent
+        // is a header union.
+        const IR::Expression* node = nullptr;
+        cstring memberName = nullptr;
+        if (auto member = expression->method->checkedTo<IR::Member>()
+                                ->expr->to<IR::Member>()) {
+            node = member->expr;
+            memberName = member->member.name;
+        } else if (auto expr = expression->method->checkedTo<IR::Member>()->expr) {
+            node = expr;
+        }
+        CHECK_NULL(node);
+        auto structVar = get(node);
         if (name == IR::Type_Header::setInvalid ||
             name == IR::Type_Header::setValid) {
-            BUG_CHECK(base->is<SymbolicHeader>(), "%1%: expected a header", base);
-            auto hv = base->to<SymbolicHeader>();
-            hv->setValid(name == IR::Type_Header::setValid);
+            if (auto hv = structVar->to<SymbolicHeader>()) {
+                hv->setValid(name == IR::Type_Header::setValid);
+            } else if (auto headerUnion = structVar->to<SymbolicHeaderUnion>()) {
```

Comment:
I don't think there is a setInvalid method call on a header_union.

Reply:
this is not the right place to evaluate the method call, it should be done in the `postorder(MethodCallExpression* )` method.

Reply:
It is the same in  `postorder(MethodCallExpression* )` . 
And this call is needed so structVar will not be passed to SymbolicHeader. And for HeadeUnion, we have implemented the setValid function in which the name of the field is passed to which setValid will be called and additionally valid will be set to other fields

Reply:
Let me repeat: there is no setValid or setInvalid call on a header_union. So this case can never happen.